nx/lib/nx.ex

```diff
@@ -1440,6 +1440,176 @@ defmodule Nx do
+  shape is `{1,2,3}` and `repetitions = [2]`, the result is the same
+  as if `repetitions = [1,1,2]`).
+
+  ## Examples
+
+      iex> a = Nx.tensor([0, 1, 2])
+      iex> Nx.tile(a, 2)
+      #Nx.Tensor<
+        s64[6]
+        [0, 1, 2, 0, 1, 2]
+      >
+      iex> Nx.tile(a, [1, 2])
+      #Nx.Tensor<
+        s64[1][6]
+        [
+          [0, 1, 2, 0, 1, 2]
+        ]
+      >
+      iex> Nx.tile(a, [2, 2])
+      #Nx.Tensor<
+        s64[2][6]
+        [
+          [0, 1, 2, 0, 1, 2],
+          [0, 1, 2, 0, 1, 2]
+        ]
+      >
+      iex> Nx.tile(a, [2, 1])
+      #Nx.Tensor<
+        s64[2][3]
+        [
+          [0, 1, 2],
+          [0, 1, 2]
+        ]
+      >
+      iex> Nx.tile(a, [2, 1, 2])
+      #Nx.Tensor<
+        s64[2][1][6]
+        [
+          [
+            [0, 1, 2, 0, 1, 2]
+          ],
+          [
+            [0, 1, 2, 0, 1, 2]
+          ]
+        ]
+      >
+
+      iex> b = Nx.tensor([[1,2],[3,4]])
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [1, 2],
+          [3, 4]
+        ]
+      >
+      iex> Nx.tile(b, 2)
+      #Nx.Tensor<
+        s64[2][4]
+        [
+          [1, 2, 1, 2],
+          [3, 4, 3, 4]
+        ]
+      >
+      iex> Nx.tile(b, [2, 1])
+      #Nx.Tensor<
+        s64[4][2]
+        [
+          [1, 2],
+          [3, 4],
+          [1, 2],
+          [3, 4]
+        ]
+      >
+      iex> Nx.tile(b, [1, 2])
+      #Nx.Tensor<
+        s64[2][4]
+        [
+          [1, 2, 1, 2],
+          [3, 4, 3, 4]
+        ]
+      >
+
+      iex> c = Nx.tensor([1,2,3,4])
+      #Nx.Tensor<
+        s64[4]
+        [1, 2, 3, 4]
+      >
+      iex> Nx.tile(c, [4,1])
+      #Nx.Tensor<
+        s64[4][4]
+        [
+          [1, 2, 3, 4],
+          [1, 2, 3, 4],
+          [1, 2, 3, 4],
+          [1, 2, 3, 4]
+        ]
+      >
+
+    ### Error cases
+
+    iex> Nx.tile(Nx.tensor([1,2]), 1.0)
+    ** (ArgumentError) repetitions must be either an integer or a list of integers. Got: 1.0
+
+    iex> Nx.tile(Nx.tensor([1,2]), [1, 1.0])
+    ** (ArgumentError) repetitions must be either an integer or a list of integers. Got: [1, 1.0]
+
+    iex> Nx.tile(Nx.tensor([1,2]), nil)
+    ** (ArgumentError) repetitions must be either an integer or a list of integers. Got: nil
+  """
+  @doc type: :shape
+  def tile(tensor, repetitions) do
+    %T{shape: old_shape} = tensor = tensor!(tensor)
+
+    invalid_repetitions_message =
+      "repetitions must be either an integer or a list of integers. Got: #{inspect(repetitions)}"
+
+    repetitions =
+      case repetitions do
+        reps when is_list(reps) ->
+          if Enum.any?(reps, &(not is_integer(&1))) do
+            raise ArgumentError, invalid_repetitions_message
+          end
+
+          reps
+
+        reps when is_integer(reps) and reps > 0 ->
+          reps
+
+        _ ->
+          raise ArgumentError, invalid_repetitions_message
+      end
+
+    repetitions = List.wrap(repetitions)
```

Comment:
If `repetitions` as an integer is the same as `[repetition]`, then I propose to not accept integer inputs. Alternatively, we can support integer inputs with a special meaning (it means repeating all dimensions with that given value).

Reply:
Yes, it's the same. I'm ok with not accepting integers. Since most of the API is akin to Numpy, having a completely different behavior for a valid input set in each one might confuse newcomers. Any thoughts?

Reply:
@josevalim aside from this discussion, the PR is ready for review :)
For now, it just rejects anything that isn't a list of integers

Reply:
>  For now, it just rejects anything that isn't a list of integers

Agreed, let's start with a restricted API. We can always add conveniences later.